In [1]:
import sys; sys.path.append('../../src/helpers')
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from data_manipulation import data
import re

## Setup

In [91]:
books = data.loadAndClean('../../data/booksummaries/booksummaries.txt')

2019-06-04 12:05:38,909 data_manipulation.data INFO     Called loadAndClean function
2019-06-04 12:05:38,914 data_manipulation.data INFO     Read in booksummary data
2019-06-04 12:05:39,611 data_manipulation.data INFO     Drop any NA's present in dataset under book genre or plot summary
2019-06-04 12:05:40,399 data_manipulation.data INFO     returning booksummary data set


In [92]:
books = books[['bookGenre', 'plotSum', 'bookTitle']]

In [93]:
books.head(1)

bookGenre  \
0  [roman_à_clef, satire, childrens_literature, speculative_fiction, fiction]   

                                                                                                                                                 plotSum  \
0   Old Major, the old boar on the Manor Farm, calls the animals on the farm for a meeting, where he compares the humans to parasites and teaches the...   

     bookTitle  
0  Animal Farm

In [97]:
books.shape[0]

12841

There are 12841 books in the data set with many genres and genre combinations. This notebook will attempt to trim down the genres to a small number (5-10) of unique categories, and possibly excise any books which don't fit into these categories.

In [95]:
corpus = list(books.bookGenre.apply(lambda row: row))
corpus[:1]

[['roman_à_clef',
  'satire',
  'childrens_literature',
  'speculative_fiction',
  'fiction']]

In [7]:
N = len(corpus)
N

12841

The corpus is the body of all of my documents, in this case a document is a list of strings which are the genre labels for each book. As you can see, the corpus has the same length as the # of books I have.

In [8]:
te = TransactionEncoder()
te_ary = te.fit(corpus).transform(corpus)
te_df = pd.DataFrame(te_ary, columns=te.columns_)

## Genre Consolidation

In [9]:
genres = list(te_df.columns)

In [10]:
len(genres)

227

The above cell shows that there are 227 unique genres which I need to pare down. The first step will be to standardize a lot of these, like turning zombies_in_pop_culture to zombie, and then the next step will be rolling things up into a smaller number of categories, like zombie --> horror.

In [29]:
books.bookGenre = corpus

In [90]:
books.head(3)

Empty DataFrame
Columns: [bookGenre, plotSum, bookTitle, Drop]
Index: []

In [22]:
te = TransactionEncoder()
te_ary = te.fit(corpus).transform(corpus)
te_df = pd.DataFrame(te_ary, columns=te.columns_)

In [23]:
frequent_itemsets = apriori(te_df, min_support=0.0001, use_colnames=True)
frequent_itemsets['Abs_support'] = frequent_itemsets.support * N

In [24]:
frequent_itemsets.head(2)

support             itemsets  Abs_support
0  0.031462                   ()        404.0
1  0.002258  (absurdist_fiction)         29.0

In [25]:
frequent_itemsets['num_of_genres'] = frequent_itemsets.itemsets.apply(lambda row: len([i for i in range(len(row))]))

In [26]:
frequent_itemsets.Abs_support.sum()

70997.0

Now we have a dataframe that we can see all the combinations and how often they occur. This will help us make the rules to classify our multi-genre labels into fewer genres. 

In [27]:
pd.set_option('display.max_colwidth', 150)
frequent_itemsets.sort_values(by='Abs_support', ascending=False)[frequent_itemsets.num_of_genres > 2].head(50)



/anaconda3/envs/AVC/lib/python3.7/site-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


support  \
671   0.092205   
605   0.078732   
561   0.052878   
629   0.045479   
447   0.044778   
604   0.040340   
1105  0.038315   
435   0.036757   
429   0.027724   
903   0.025621   
651   0.021416   
468   0.021026   
446   0.019080   
677   0.018612   
559   0.017678   
939   0.017444   
451   0.016354   
584   0.012460   
617   0.011993   
1072  0.011526   
600   0.011058   
476   0.011058   
679   0.010980   
434   0.010435   
915   0.009968   
1095  0.009734   
902   0.009579   
1284  0.009111   
420   0.008800   
698   0.008800   
947   0.008021   
637   0.008021   
449   0.008021   
555   0.007554   
459   0.007398   
644   0.007242   
650   0.007009   
440   0.007009   
1151  0.006853   
441   0.006697   
717   0.006386   
702   0.006308   
550   0.006308   
570   0.006308   
673   0.006230   
609   0.006152   
930   0.006074   
653   0.005996   
544   0.005996   
715   0.005919   

                                                                  itemsets  \
671                        (science_fiction, speculative_fiction, fiction)   
605                                (speculative_fiction, fiction, fantasy)   
561                        (crime_fiction, fiction, suspense/thriller/spy)   
629                        (science_fiction, speculative_fiction, fantasy)   
447                               (speculative_fiction, fiction, children)   
604                                    (science_fiction, fiction, fantasy)   
1105              (science_fiction, speculative_fiction, fiction, fantasy)   
435                               (speculative_fiction, fantasy, children)   
429                                           (fiction, fantasy, children)   
903                      (speculative_fiction, fiction, fantasy, children)   
651                                 (horror, speculative_fiction, fiction)   
468                       (science_fiction, speculative_fiction, children)   
446                                   (science_fiction, fiction, children)   
677                  (speculative_fiction, fiction, suspense/thriller/spy)   
559                          (speculative_fiction, crime_fiction, fiction)   
939              (science_fiction, speculative_fiction, fiction, children)   
451                                       (young_adult, fiction, children)   
584            (speculative_fiction, crime_fiction, suspense/thriller/spy)   
617                                 (horror, speculative_fiction, fantasy)   
1072  (speculative_fiction, crime_fiction, fiction, suspense/thriller/spy)   
600                                             (horror, fiction, fantasy)   
476                           (speculative_fiction, young_adult, children)   
679                            (speculative_fiction, young_adult, fiction)   
434                                   (science_fiction, fantasy, children)   
915              (science_fiction, speculative_fiction, fantasy, children)   
1095                       (horror, speculative_fiction, fiction, fantasy)   
902                          (science_fiction, fiction, fantasy, children)   
1284    (speculative_fiction, fantasy, science_fiction, children, fiction)   
420                                     (crime_fiction, fiction, children)   
698                         (science_fiction, speculative_fiction, horror)   
947                  (speculative_fiction, young_adult, fiction, children)   
637                            (speculative_fiction, young_adult, fantasy)   
449                             (suspense/thriller/spy, fiction, children)   
555                                       (horror, crime_fiction, fiction)   
459                                (speculative_fiction, horror, children)   
644                     (speculative_fiction, historical_fiction, fiction)   
650                                     (horror, science_fiction, fiction)   
440                                (historical_fiction, fiction, children)   
1151               (horror, speculative_fictio